### PR TITLE
ArgumentStash for Scalar arguments

### PR DESCRIPTION
**Pull request by lara-hdr**
Scalars are being traced as constants.
This PR is to fix this issue.

The ONNX Graph for Test_Full_op() before and after this change:

def Test_Full_op():
  class Test_Full(nn.Module):
    def forward(self, x):
      return torch.full((3, 4), x, dtype=torch.long)
  model = Test_Full()
  x = torch.tensor(12)
  output = model(x)

Before this change:
graph(%input1 : Long()):
%output1 : Float(3, 4) = onnx::Constant[value=<Tensor>]
return (%output1)

After this change:
graph(%input1 : Long()):
%1 : int[] = onnx::Constant[value= 3 4 [ Variable[CPULongType]{2} ]]
%2 : Tensor = onnx::ConstantOfShape[value={0}]
%output1 : Float(3, 4) = onnx::Add(%2, %input1)
return (%output1)

Similar PR : https://github.com/pytorch/pytorch/pull/12939

### CODE REVIEW

**Review by houseroad — COMMENTED** [review]
Also a test case in caffe2 backend?

**Comment by soumith** [comment]
@houseroad looks like the internal diff can be shipped now, tests are all clear

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -6,13 +6,46 @@
...
+import numpy as np
+import io
 
 class TestONNXRuntime(unittest.TestCase):
 
     def test_onnxruntime_installed(self):
         self.assertTrue('onnxruntime' in sys.modules)
 
+    def test_full(self):
+        class FullModel(torch.nn.Module):
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: As I mentioned in another PR, could you refactor the test case bit by creating some helper function.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -6,13 +6,46 @@
...
+        output = model(x)
+
+        f = io.BytesIO()
+        torch.onnx.export(model, x, f)
+
+        ort_sess = onnxruntime.InferenceSession(f.getvalue())
+        ort_outs = ort_sess.run(None, {ort_sess.get_inputs()[0].name: x.numpy()})
+        np.allclose(output.numpy(), ort_outs[0])
+
+        class FullModelScripting(torch.jit.ScriptModule):
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: Also you can create this test case into 2 small case, test_full_trace and test_full_script.

File: torch/csrc/utils/python_arg_parser.h [hunk 3]
```diff
@@ -236,6 +236,12 @@ inline at::Tensor PythonArgs::tensor(int i) {
 }
 
 inline at::Scalar PythonArgs::scalar(int i) {
+  if (!args[i]) return signature.params[i].default_scalar;
+  if (traceable && jit::tracer::isTracing() && THPVariable_Check(args[i])) {
+    auto & var = THPVariable_Unpack(args[i]);
```
Comment (houseroad) [inline torch/csrc/utils/python_arg_parser.h]: auto & ==> auto&

**Review by houseroad — COMMENTED** [review]
Overall, looks good. Please address the comments.

**Review by houseroad — APPROVED** [review]
LGTM.

**Pull request closed without merging**